include/crocoddyl/multibody/actions/contact-invdyn.hpp

```diff
@@ -0,0 +1,618 @@
+      df_du.block(fid, nv + fid, nc, nc).diagonal().setOnes();
+      fid += nc;
+    }
+    model->get_contacts()->updateForceDiff(multibody.contacts, df_dx, df_du);
+    costs = model->get_costs()->createData(&multibody);
+    constraints = model->get_constraints()->createData(&multibody);
+    costs->shareMemory(this);
+    constraints->shareMemory(this);
+
+    tmp_xstatic.setZero();
+    tmp_rstatic.setZero();
+    tmp_Jstatic.setZero();
+  }
+
+  pinocchio::DataTpl<Scalar> pinocchio;                  //!< Pinocchio data
+  DataCollectorJointActMultibodyInContact multibody;     //!< Multibody data
+  boost::shared_ptr<CostDataSum> costs;                  //!< Costs data
+  boost::shared_ptr<ConstraintDataManager> constraints;  //!< Constraints data
+  VectorXs tmp_xstatic;                                  //!< State point used for computing the quasi-static input
+  VectorXs tmp_rstatic;                                  //!< Factorization used for computing the quasi-static input
+  MatrixXs tmp_Jstatic;                                  //!< Jacobian used for computing the quasi-static input
+
+  using Base::cost;
+  using Base::Fu;
+  using Base::Fx;
+  using Base::Lu;
+  using Base::Luu;
+  using Base::Lx;
+  using Base::Lxu;
+  using Base::Lxx;
+  using Base::r;
+  using Base::xout;
+
+  struct ResidualDataActuation : public ResidualDataAbstractTpl<_Scalar> {
+    EIGEN_MAKE_ALIGNED_OPERATOR_NEW
+
+    typedef _Scalar Scalar;
+    typedef MathBaseTpl<Scalar> MathBase;
+    typedef ResidualDataAbstractTpl<Scalar> Base;
+    typedef DataCollectorAbstractTpl<Scalar> DataCollectorAbstract;
+    typedef DataCollectorActMultibodyInContactTpl<Scalar> DataCollectorActMultibodyInContact;
+    typedef ActuationDataAbstractTpl<Scalar> ActuationDataAbstract;
+    typedef ContactDataMultipleTpl<Scalar> ContactDataMultiple;
+    typedef typename MathBase::MatrixXs MatrixXs;
+
+    template <template <typename Scalar> class Model>
+    ResidualDataActuation(Model<Scalar>* const model, DataCollectorAbstract* const data)
+        : Base(model, data),
+          dtau_dx(model->get_state()->get_nv(), model->get_state()->get_ndx()),
+          dtau_du(model->get_state()->get_nv(), model->get_nu()) {
+      // Check that proper shared data has been passed
+      DataCollectorActMultibodyInContact* d = dynamic_cast<DataCollectorActMultibodyInContact*>(shared);
+      if (d == NULL) {
+        throw_pretty("Invalid argument: the shared data should be derived from DataCollectorActMultibodyInContact");
+      }
+      // Avoids data casting at runtime
+      pinocchio = d->pinocchio;
+      actuation = d->actuation.get();
+      contact = d->contacts.get();
+      dtau_dx.setZero();
+      dtau_du.setZero();
+    }
+
+    pinocchio::DataTpl<Scalar>* pinocchio;  //!< Pinocchio data
+    ActuationDataAbstract* actuation;       //!< Actuation data
+    ContactDataMultiple* contact;           //!< Contact data
```

Comment:
Would it make sense to use a smart pointer here?

Reply:
I wanted to avoid create another owner, but I guess it makes sense to use a smart pointer for safety. I will do it.

Reply:
You could use a `weak_ptr` to avoid having an owner; in that case you'd need to `my_weak_ptr.lock()` every time to create a `shared_ptr` (which has some overhead). There might be better options though?

Reply:
Thanks for the advise, I will stick with shared pointers (at least) for now.